### PR TITLE
bpo-32984: IDLE - set __file__ for startup files

### PR DESCRIPTION
**Pull request by terryjreedy**
https://bugs.python.org/issue32984

### CODE REVIEW

File: Lib/idlelib/pyshell.py [hunk 1]
```diff
@@ -635,6 +635,9 @@ def execfile(self, filename, source=None):
         if source is None:
             with tokenize.open(filename) as fp:
                 source = fp.read()
+                if use_subprocess:
+                    source = (f"__file__ = r'{os.path.abspath(filename)}'\n" +
+                              source + "\ndel __file__")
```
Comment (csabella) [inline Lib/idlelib/pyshell.py]: This gives a SyntaxError if there's an unescaped quote in the path or file name (such as `/home/cheryl's/tests/`).

Using triple quotes works:
```f"__file__ = r'''{os.path.abspath(filename)}'''\n"```

File: Misc/NEWS.d/next/IDLE/2018-03-05-01-29-05.bpo-32984.NGjgT4.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Set __file__ while running a startup file. Like Python, IDLE optionally runs
+1 startup file in the Shell window before presenting the first interactive
+input prompt.  For IDLE, -s runs a file named in environmental variable
+IDLESTARTUP or PYTHONSTARTUP; -r file runs file.  Python sets __file__ to
+the startup file name before running the file and unsets it before the first
+prompt.  IDLE now does the same when run normally, without the -n option.
```
Comment (csabella) [inline Misc/NEWS.d/next/IDLE/2018-03-05-01-29-05.bpo-32984.NGjgT4.rst]: Set ``__file__`` while running a startup file.  Like Python, IDLE optionally
runs one startup file in the Shell window before presenting the first
interactive input prompt.  For IDLE, ``-s`` runs a file named in
environmental variable :envvar:`IDLESTARTUP` or :envvar:`PYTHONSTARTUP`;
``-r`` file runs file.  Python sets ``__file__`` to the startup file
name before running the file and unsets it before the first prompt.  IDLE now
does the same when run normally, without the ``-n`` option.

Comment (terryjreedy) [reply Misc/NEWS.d/next/IDLE/2018-03-05-01-29-05.bpo-32984.NGjgT4.rst]: Thanks.  by 'editing', I could see and copy the markup (and maybe learn better).  I changed one clause to ``-r file`` runs ``file``,  which I can preview here ;-).

File: Lib/idlelib/pyshell.py (hunk 1 above)
Comment (terryjreedy) [reply Lib/idlelib/pyshell.py]: Great catch.  Changed.

**Pull request merged**